test/es-module/test-esm-import-attributes-identity.mjs

```diff
@@ -0,0 +1,52 @@
+import '../common/index.mjs';
+import assert from 'node:assert';
+import Module from 'node:module';
+
+// This test verifies that importing two modules with different import
+// attributes should result in different module instances, if the module loader
+// resolves to module instances.
+//
+// For example,
+// ```mjs
+//   import * as secret1 from '../primitive-42.json' with { type: 'json' };
+//   import * as secret2 from '../primitive-42.json';
+// ```
+// should result in two different module instances, if the second import
+// is been evaluated as a CommonJS module.
+//
+// ECMA-262 requires that in `HostLoadImportedModule`, if the operation is called
+// multiple times with two (referrer, moduleRequest) pairs, it should return the same
+// result. But if the moduleRequest is different, it should return a different,
+// and the module loader resolves to different module instances, it should return
+// different module instances.
+// Refs: https://tc39.es/ecma262/#sec-HostLoadImportedModule
+
+const kJsonModuleSpecifier = '../primitive-42.json';
+const fixtureUrl = new URL('../fixtures/primitive-42.json', import.meta.url).href;
+Module.registerHooks({
+  resolve: (specifier, context, nextResolve) => {
+    console.log('resolve', specifier, context);
+    if (kJsonModuleSpecifier !== specifier) {
+      return nextResolve(specifier, context);
+    }
+    if (context.importAttributes.type === 'json') {
+      return {
+        url: fixtureUrl,
+        importAttributes: context.attributes,
+        shortCircuit: true,
+        format: 'json',
+      };
+    }
+    return {
+      url: fixtureUrl,
+      importAttributes: context.attributes,
+      shortCircuit: true,
```

Comment:
It would be nice to see a test where if the attributes end up being altered in the resolve hook to be the same attributes (i.e. the import attribute "canonicalization"), that the module instance is then the same.

Would also be great to see the interaction with source phase imports where identity is maintained under the same principles, but that can be a follow-up.

Reply:
Added a test on a module loader returning a new import attributes object.